### PR TITLE
Remove the legacy query planner

### PR DESCRIPTION
**Pull request by SimonSapin**
The legacy query planner has been removed in this release. In the previous release, Router 1.58, it was already no longer used by default but it was still available through the `experimental_query_planner_mode`  configuration key. That key is now removed.

Also removed are configuration keys which were only relevant to the legacy planner:

* `supergraph.query_planning.experimental_parallelism`: the new planner can always use available parallelism.
* `supergraph.experimental_reuse_query_fragments`: this experimental algorithm that attempted to
reuse fragments from the original operation while forming subgraph requests is no longer present. Instead, by default new fragment definitions are generated based on the shape of the subgraph operation.

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

**Comment by samuelAndalon** [comment]
image

**Comment by SimonSapin** [comment]
This is ready for review but should only be merged after we branch 1.59 later this week.

File: apollo-router/src/configuration/migrations/0032-remove-legacy-qp.yaml
Comment (lrlna) [inline apollo-router/src/configuration/migrations/0032-remove-legacy-qp.yaml:1]: oh we need another migration for removing `experimental_query_planner_mode`

**Review by lrlna — COMMENTED** [review]
Potentially the most exciting PR in router's history! 🎉 

Leaving some initial comments.

File: apollo-router/tests/integration/query_planner.rs [hunk 1]
```diff
@@ -61,112 +25,10 @@ async fn fed1_schema_with_both_qp() {
...
-    router.execute_default_query().await;
-    router.graceful_shutdown().await;
-}
-
 #[tokio::test(flavor = "multi_thread")]
 async fn fed2_schema_with_new_qp() {
-    let config = format!("{PROMETHEUS_METRICS_CONFIG}\n{NEW_QP}");
     let mut router = IntegrationTest::builder()
-        .config(config)
+        .config(PROMETHEUS_METRICS_CONFIG)
```
[the remark this answers is not in the document]
Comment (SimonSapin) [reply apollo-router/tests/integration/query_planner.rs:31]: The `apollo.router.lifecycle.query_planner.init` metric still exists and is tested here. Should we also remove it?

File: apollo-router/tests/integration/redis.rs [hunk 2]
```diff
@@ -992,7 +991,7 @@ async fn query_planner_redis_update_query_fragments() {
         // This configuration turns the fragment generation option *off*.
         include_str!("fixtures/query_planner_redis_config_update_query_fragments.router.yaml"),
         &format!(
-            "plan:router:{}:5938623f2155169070684a48be1e0b8468d0f2c662b5527a2247f683173f7d05:opname:3973e022e93220f9212c18d0d0c543ae7c309e46640da93a4a0314de999f5112:metadata:28acec2bebc3922cd261ed3c8a13b26d53b49e891797a199e3e1ce8089e813e6",
+            "plan:router:{}:5938623f2155169070684a48be1e0b8468d0f2c662b5527a2247f683173f7d05:opname:3973e022e93220f9212c18d0d0c543ae7c309e46640da93a4a0314de999f5112:metadata:fb1a8e6e454ad6a1d0d48b24dc9c7c4dd6d9bf58b6fdaf43cd24eb77fbbb3a17",
             env!("CARGO_PKG_VERSION")
         ),
     )
```
[the remark this answers is not in the document]
Comment (SimonSapin) [reply apollo-router/tests/integration/redis.rs:997]: Done

File: apollo-router/src/configuration/migrations/0032-remove-legacy-qp.yaml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+description: the legacy query planner was removed, together with relevant configuration keys experimental_query_planner_mode and supergraph.query_planning.experimental_parallelism
+actions:
+  - type: delete
```
[the remark this answers is not in the document]
Comment (SimonSapin) [reply apollo-router/src/configuration/migrations/0032-remove-legacy-qp.yaml:3]: Done

File: apollo-router/src/query_planner/plan_compare.rs
[the remark this answers is not in the document]
Comment (SimonSapin) [reply apollo-router/src/query_planner/plan_compare.rs:1]: Do we plan on doing that? I think if and when that happens we can find the code in git history. Or is it worth to keep maintaining it even if unused in the mean time?

File: apollo-router/src/query_planner/bridge_query_planner.rs
[the remark this answers is not in the document]
Comment (lrlna) [reply apollo-router/src/query_planner/bridge_query_planner.rs:1]: ah, follow ups are fine then.

File: apollo-router/src/query_planner/plan_compare.rs
[the remark this answers is not in the document]
Comment (lrlna) [reply apollo-router/src/query_planner/plan_compare.rs:1]: No, not really worth it, I guess. 

@duckki are any of the tools we have that do semantic equivalence rely on the router's version of it directly?

[the remark this answers is not in the document]
Comment (duckki) [reply apollo-router/src/query_planner/plan_compare.rs:1]: `federation-performance-harness` is the only one, I think. `plan compare` command would be no longer needed. Still, we can have a separate branch that uses the current version of router and compares query plans.

Also, I have cloned semantic diff in federation-test-lab ([private github](https://github.com/mdg-private/federation-test-lab)) and keep them in sync. We can still use it for qp comparison without router.

[the remark this answers is not in the document]
Comment (SimonSapin) [reply apollo-router/src/query_planner/plan_compare.rs:1]: FYE, when merging from dev today I got conflicts with https://github.com/apollographql/router/pull/6431 in `plan_compare.rs` removed by this PR. But it looks you’ve already taken care of having these changes `federation-test-lab`, so I think we’re good.

**Review by garypen — APPROVED** [review]
This looks great! We should follow up at some point with some renaming of `BridgeQueryPlanner` bits and I think we could completely delete `BridgeQueryPlannerPool`.

**Pull request merged**